pkg/controller/core/resource_controller.go

```diff
@@ -156,16 +160,21 @@ func (r *Reconciler) _provision(claim corev1alpha1.ResourceClaim, handler Resour
 	}
 
 	// add finalizer
-	util.AddFinalizer(claim, r.finalizerName)
+	meta.AddFinalizer(claim, r.finalizerName)
 
 	// create new resource
-	res, err := handler.Provision(class, claim, r.Client)
+	resource, err := handler.Provision(class, claim, r.Client)
 	if err != nil {
 		return r.fail(claim, errorResourceProvisioning, err.Error())
 	}
 
+	gvk, err := apiutil.GVKForObject(resource, r.scheme)
```

Comment:
does this make a network call to the API server? or does this use the client side scheme information only?

in this generic resource controller, we've moved from a model where we could ask each resource about its type information to a model where we have to do some reflectionish behavior here to understand what type a resource is.  In the previous approach, that resulted in all the types having to implement the `OwnerReference()` and `ObjectReference()` functions.

Now, those are no longer part of the interface and they don't have to be implemented over and over again, but it has the cost of having potentially less confidence or understanding what the type of a resource is at runtime.  Take for example how this line here is a new error path that didn't exist before.

Similarly, every call to `meta.ReferenceTo` now has to know the type of the resource it's asking for a reference to, instead of the resource being able to provide that itself.

With everything you've seen while implementing this, do you think that's a good trade-off?  I trust your response, but wanted to ask explicitly about this choice.  Thanks!

Reply:
> does this make a network call to the API server? or does this use the client side scheme information only?

It uses client side scheme information only. To my knowledge there's no code in `*runtime.Scheme` that makes network calls, and [`apiutil.GVKForObject`](https://github.com/kubernetes-sigs/controller-runtime/blob/f99287c/pkg/client/apiutil/apimachinery.go#L47) is a simple wrapper around `s: = *runtime.NewScheme(); s.ObjectKinds(obj)` that handles some of its edge cases.


> Every call to meta.ReferenceTo now has to know the type of the resource it's asking for a reference to, instead of the resource being able to provide that itself.

A large part of the reason I took this path is that resources _could already_ provide that type data using their `TypeMeta` regardless of our `ObjectReference` and `OwnerReference` convenience methods, but as best I can tell it is (was?) considered bad practice to associate API type metadata with Go structs. Objects actively have their type metadata cleared during deserialisation. There's some context in https://github.com/kubernetes-sigs/controller-runtime/issues/406. It was at some point intentional; it's unclear to me whether it still is.

> Do you think that's a good trade-off?

It's less of an obvious win than my initial pass given we can't rely on our types having valid `TypeMeta`, but I still personally prefer the flexibility of a method that can be called on a type to requiring types implement convenience methods. I also like that this method has symmetry with [similar upstream methods](https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#NewControllerRef), and that it reduces the surface area required to satisfy the `Resource` interface by removing methods that provide data that can almost always be pretty easily found elsewhere.

Some examples of using this API.
```go
rc := &v1alpha1.RedisCluster{}

// In almost all cases GVK can be provided explicitly. This is close to the
// contemporary implementation, with the exception that the caller must
// pass in the static GVK. This satisfies the 
ref := meta.ReferenceTo(rc, v1alpha1.RedisClusterGroupVersionKind)

// If the object is known to have type metadata we can use it.
if gvk := rc.ObjectKind().GroupVersionKind(); !gvk.Empty() {
    ref = meta.ReferenceTo(rc, gvk)
}

// Worst case, as in the core resource controller, the object is wrapped up
// in an interface and contains no type metadata. In this case we rely on the
// fact that all resource types we're capable of handling are registered with
// the scheme. 
ref = meta.ReferenceTo(rc, MustGetKind(rc, mgr.GetScheme()))

func MustGetKind(o runtime.Object, s *runtime.Scheme) schema.GroupVersionKind {
	gvk, err := apiutil.GVKForObject(o, s)
	if err != nil {
		panic(err.Error())
	}
	return gvk
}
```

Reply:
We might consider adding the above `MustGetKind` method. I believe that with the exception of our controllers that work with `*unstructured.Unstructured` it's safe for us to assume that any object being reconciled by a controller is known to `mgr.GetScheme()`.

Reply:
FYI, I added `MustGetKind` and converted the core resource claim controller to use it. I tested it by deploying a GCP workload, which includes creating a KubernetesCluster and MySQLInstance. As expected, nothing paniced. :)

Reply:
Thank you for the detailed response on this @negz.  I think this is a decent approach.